client/object_put.go

```diff
@@ -45,6 +45,15 @@ type PrmObjectPutInit struct {
 	sessionContainer
 
 	copyNum uint32
+
+	buf []byte
+}
+
+// SetBuf allows to set pre-allocated buffer for object uploading operation.
+//
+// For better performance buf size should be equal to MaxObjectSize from NeoFS.
+func (x *PrmObjectPutInit) SetBuf(buf []byte) {
```

Comment:
Can we just have a client-level `sync.Pool` for gRPC messages (max message size is known)? It'd reuse it for _any_ messages, not just PUT-related ones. And this would allow to avoid additional parameters. Also, gRPC messages are different from object payload slices in their size.

Reply:
I consider it is possible. Is the required buffer size 4mb? Almost all places work good for this change, but PUT will be complicated because we pass our buffer to ObjectWritter and return it to outside. In this case, we don't know when we should _return_ the preallocated buffer - we need some extra logic inside ObjectWritter to _signal_ that we may return the buffer to the pool.

I will try and show the result.

Reply:
Reworked

Reply:
Benched?

Reply:
Not yet, I will handle it